src/main/kotlin/teaforge/platform/RoboRio/internal/Subscriptions.kt

```diff
@@ -206,3 +228,25 @@ fun <TMessage, TModel> runReadHidPort(
 
         return Triple(model, state, Maybe.Some(state.config.message(hidValue)))
 }
+
+fun <TMessage, TModel> runReadInterval(
+        model: RoboRioModel<TMessage, TModel>,
+        state: SubscriptionState.Interval<TMessage>
+): Triple<RoboRioModel<TMessage, TModel>, SubscriptionState<TMessage>, Maybe<TMessage>>{
+        val currentMicroseconds = HALUtil.getFPGATime()
+        return if (currentMicroseconds >= state.nextReadTimeMicroseconds){
+                val intervalMicroseconds = state.config.millisecondsBetweenReads * 1_000L
+                val intervalsMissed = ((currentMicroseconds - state.nextReadTimeMicroseconds) / intervalMicroseconds) + 1
+
+                val newNextReadTime = state.nextReadTimeMicroseconds + intervalMicroseconds * intervalsMissed
+
+                val elapsedMilliseconds = (currentMicroseconds - state.nextReadTimeMicroseconds) / 1_000L
```

Comment:
If you move this earlier you could reuse it for the logic of `intervalsMissed`. Also I think it would be good to stay in the microseconds unit and convert to milliseconds at the very end.

```suggestion
                val elapsedMicroseconds = currentMicroseconds - state.nextReadTimeMicroseconds
```

Then you can do

```
val intervalsMissed = elapsedMicroseconds / intervalMicroseconds
```

I see you were adding +1 here to get to the next interval but that represents the *next* interval, not the number missed. So when you go to calculate the updated `newNextReadTime` that's the time to incorporate the +1

```
val newNextReadTime = state.nextReadTimeMicroseconds + (intervalMicroseconds * (intervalsMissed + 1))
```